langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/utils/SortUtils.java

```diff
@@ -0,0 +1,146 @@
+                        return false;
+                    }
+                }
+                return true;
+            case TypeTags.ARRAY_TAG:
+                return isOrderedType(((BArrayType) type).getElementType());
+            case TypeTags.TUPLE_TAG:
+                List<Type> tupleTypes = ((BTupleType) type).getTupleTypes();
+                if (((BTupleType) type).getRestType() != null) {
+                    tupleTypes.add(((BTupleType) type).getRestType());
+                }
+                if (!isOrderedType(tupleTypes.get(0))) {
+                    return false;
+                }
+                for (Type memType : tupleTypes) {
+                    if (!isOrderedType(memType) || isDifferentOrderedType(memType, tupleTypes.get(0))) {
+                        return false;
+                    }
+                }
+                return true;
+            case TypeTags.FINITE_TYPE_TAG:
+                Set<Object> valSpace = ((BFiniteType) type).getValueSpace();
+                Type baseExprType = TypeUtils.getType(valSpace.iterator().next());
+                if (!checkValueSpaceHasSameType((BFiniteType) type, baseExprType)) {
+                    return false;
+                }
+                return isOrderedType(baseExprType);
+            default:
+                return isSimpleBasicType(type.getTag());
+        }
+    }
+
+    /**
+     * Check if the values space of the provided finite type belongs to the value space of the given type.
+     * @param finiteType finite type to be checked.
+     * @param type type to be checked against.
+     * @return true if the finite type belongs to the same value space, false otherwise.
+     */
+    public static boolean checkValueSpaceHasSameType(BFiniteType finiteType, Type type) {
+        Type baseType = TypeUtils.getImpliedType(type);
+        if (baseType.getTag() == TypeTags.FINITE_TYPE_TAG) {
+            return checkValueSpaceHasSameType((BFiniteType) baseType,
+                    TypeUtils.getType(finiteType.getValueSpace().iterator().next()));
+        }
+        for (Object expr : finiteType.getValueSpace()) {
+            if (isDifferentOrderedType(TypeUtils.getType(expr), baseType)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Check whether a given type is different to a target type.
+     * @param source type to check.
+     * @param target type to compare with.
+     * @return true if the source type does not belong to the target type, false otherwise.
+     */
+    public static boolean isDifferentOrderedType(Type source, Type target) {
+        source = TypeUtils.getImpliedType(source);
+        target = TypeUtils.getImpliedType(target);
+        if (source.getTag() == TypeTags.NULL_TAG || target.getTag() == TypeTags.NULL_TAG) {
+            return false;
+        }
+        return !TypeChecker.checkIsType(source, target);
```

Comment:
Do we handle the `refered` type in this logic

Reply:
Since `TypeUtils.getImpliedType()` returns the `referred` or the `effective` type appropriately, both scenarios are covered right?